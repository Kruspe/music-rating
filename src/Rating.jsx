import React, { useState } from 'react';
import { Rating as RatingMaterialUI } from '@material-ui/lab';

const Rating = () => {
  const [rating, setRating] = useState(0);
  return (
    <RatingMaterialUI value={rating} onChange={(event, value) => setRating(value)} />
  );
};

export default Rating;
