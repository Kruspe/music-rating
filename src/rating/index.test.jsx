import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { Rating as RatingMaterialUI } from '@material-ui/lab';
import { API, Auth } from 'aws-amplify';
import Rating from './index';

let wrapper;
describe('Rating', () => {
  beforeEach(() => {
    wrapper = shallow(<Rating />);
    jest.clearAllMocks();
  });
  it('should render correctly', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  describe('form', () => {
    it('should sent data to api and should call event.preventDefault', async () => {
      const preventDefaultMock = jest.fn();
      const currentSessionMock = Promise.resolve({
        getAccessToken: () => ({ getJwtToken: () => ('Token') }),
      });
      const expectedInit = {
        header: { Authorization: 'Bearer Token' },
        body: { bandName: 'bandName', rating: 4, comment: 'comment' },
      };
      const postSpy = jest.spyOn(API, 'post').mockImplementation(f => f);
      jest.spyOn(Auth, 'currentSession').mockImplementation(() => currentSessionMock);
      wrapper.find('#bandName').prop('onChange')({ target: { value: 'bandName' } });
      wrapper.find(RatingMaterialUI).prop('onChange')({}, 4);
      wrapper.find('#comment').prop('onChange')({ target: { value: 'comment' } });

      await wrapper.find('#rating-form').prop('onSubmit')({ preventDefault: preventDefaultMock });
      expect(preventDefaultMock).toHaveBeenCalled();
      expect(postSpy).toHaveBeenCalledWith('musicrating', '/bands', expectedInit);
    });

    it('should not submit data if bandName is not filled', async () => {
      const preventDefaultMock = jest.fn();
      const postSpy = jest.spyOn(API, 'post').mockImplementation(f => f);
      wrapper.find('#bandName').prop('onChange')({ target: { value: '' } });
      await wrapper.find('#rating-form').prop('onSubmit')({ preventDefault: preventDefaultMock });
      wrapper.find('#bandName').prop('onChange')({ target: { value: '  ' } });
      await wrapper.find('#rating-form').prop('onSubmit')({ preventDefault: preventDefaultMock });
      wrapper.find('#bandName').prop('onChange')({ target: { value: undefined } });
      await wrapper.find('#rating-form').prop('onSubmit')({ preventDefault: preventDefaultMock });
      expect(preventDefaultMock).toHaveBeenCalledTimes(3);
      expect(postSpy).not.toHaveBeenCalled();
    });
  });

  describe('bandName', () => {
    it('should have empty string as initial value', () => {
      expect(wrapper.find('#bandName').prop('value')).toBe('');
    });
    it('should update TextField onChange', () => {
      wrapper.find('#bandName').prop('onChange')({ target: { value: 'bandName' } });
      expect(wrapper.find('#bandName').prop('value')).toBe('bandName');
    });
  });

  describe('rating', () => {
    it('should have 0 as initial state', () => {
      expect(wrapper.find(RatingMaterialUI).prop('value')).toBe(1);
    });
    it('should update rating onChange', () => {
      wrapper.find(RatingMaterialUI).prop('onChange')({}, 2);
      expect(wrapper.find(RatingMaterialUI).prop('value')).toBe(2);
    });
  });

  describe('comment', () => {
    it('should have empty string as initial state', () => {
      expect(wrapper.find('#comment').prop('value')).toBe('');
    });
    it('should update rating onChange', () => {
      wrapper.find('#comment').prop('onChange')({ target: { value: 'comment' } });
      expect(wrapper.find('#comment').prop('value')).toBe('comment');
    });
  });
});
