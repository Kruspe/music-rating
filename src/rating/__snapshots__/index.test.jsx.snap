// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating should render correctly 1`] = `
<form
  id="rating-form"
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    className="rating-container"
    container={true}
    justify="center"
    spacing={5}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={2}
    >
      <WithStyles(ForwardRef(TextField))
        id="bandName"
        label="Band Name"
        onChange={[Function]}
        value=""
        variant="outlined"
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="rating-rating"
      item={true}
    >
      <WithStyles(ForwardRef(Rating))
        name="rating"
        onChange={[Function]}
        value={1}
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={4}
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="comment"
        label="Comment"
        onChange={[Function]}
        value=""
        variant="outlined"
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={1}
    >
      <WithStyles(ForwardRef(Button))
        type="submit"
        variant="outlined"
      >
         Submit 
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</form>
`;
